File: codes/driven_cavity/driven_cavity_2d_staggered.m
% ----------------------------------------------------------------------- %
%                                     __  __  __       _  __   __         %
%        |\/|  _  |_ |  _  |_   |__| /   |_  |  \  _  (_ |__) |_          %
%        |  | (_| |_ | (_| |_)     | \__ |   |__/ (_) |  | \  |           %
%                                                                         %
% ----------------------------------------------------------------------- %
%                                                                         %
%   Author: Alberto Cuoci <alberto.cuoci@polimi.it>                       %
%   CRECK Modeling Group <http://creckmodeling.chem.polimi.it>            %
%   Department of Chemistry, Materials and Chemical Engineering           %
%   Politecnico di Milano                                                 %
%   P.zza Leonardo da Vinci 32, 20133 Milano                              %
%                                                                         %
% ----------------------------------------------------------------------- %
%                                                                         %
%   This file is part of Matlab4CFDofRF framework.                        %
%                                                                         %
%   License                                                               %
%                                                                         %
%   Copyright(C) 2019 Alberto Cuoci                                       %
%   Matlab4CFDofRF is free software: you can redistribute it and/or       %
%   modify it under the terms of the GNU General Public License as        %
%   published by the Free Software Foundation, either version 3 of the    %
%   License, or (at your option) any later version.                       %
%                                                                         %
%   Matlab4CFDofRF is distributed in the hope that it will be useful,     %
%   but WITHOUT ANY WARRANTY; without even the implied warranty of        %
%   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         %
%   GNU General Public License for more details.                          %
%                                                                         %
%   You should have received a copy of the GNU General Public License     %
%   along with Matlab4CRE. If not, see <http://www.gnu.org/licenses/>.    %
%                                                                         %
%-------------------------------------------------------------------------%
%                                                                         %
%  Code: 2D driven-cavity problem on a staggered grid                     %
%        The code is adapted and extended from Tryggvason, Computational  %
%        Fluid Dynamics http://www.nd.edu/~gtryggva/CFD-Course/           %
%                                                                         %
% ----------------------------------------------------------------------- %
close all;
clear variables;

% ----------------------------------------------------------------------- %
% User data
% ----------------------------------------------------------------------- %

% Only even numbers of cells are acceptable
nx=24;      % number of (physical) cells along x
ny=nx;      % number of (physical) cells along y
L=1;        % length [m]
nu=1e-2;    % kinematic viscosity [m2/s] (if L=1 and un=1, then Re=1/nu)
tau=20;     % total time of simulation [s]

% Boundary conditions
un=1;       % north wall velocity [m/s]
us=0;       % south wall velocity [m/s]
ve=0;       % east wall velocity [m/s]
vw=0;       % west wall velocity [m/s]

% Parameters for SOR
max_iterations=10000;   % maximum number of iterations
beta=1.5;               % SOR coefficient
max_error=1e-4;         % error for convergence

% ----------------------------------------------------------------------- %
% Data processing
% ----------------------------------------------------------------------- %

% Grid step
h=L/nx;                             % grid step (uniform grid) [m]

% Time step
sigma = 0.5;                        % safety factor for time step (stability)
dt_diff=h^2/4/nu;                   % time step (diffusion stability) [s]
dt_conv=4*nu/un^2;                  % time step (convection stability) [s]
dt=sigma*min(dt_diff, dt_conv);     % time step (stability) [s]
nsteps=tau/dt;                      % number of steps
Re = un*L/nu;                       % Reynolds' number

fprintf('Time step: %f\n', dt);
fprintf(' - Diffusion:  %f\n', dt_diff);
fprintf(' - Convection: %f\n', dt_conv);
fprintf('Reynolds number: %f\n', Re);

% Grid construction
x=0:h:1;                % grid coordinates (x axis)
y=0:h:1;                % grid coordinates (y axis)
[X,Y] = meshgrid(x,y);  % MATLAB grid

% ----------------------------------------------------------------------- %
% Memory allocation
% ----------------------------------------------------------------------- %

% Main fields (velocities and pressure)
u=zeros(nx+1,ny+2);
v=zeros(nx+2,ny+1);
p=zeros(nx+2,ny+2);

% Temporary velocity fields
ut=zeros(nx+1,ny+2);
vt=zeros(nx+2,ny+1);

% Fields used only for graphical post-processing purposes
uu=zeros(nx+1,ny+1);
vv=zeros(nx+1,ny+1);
pp=zeros(nx+1,ny+1);

% Coefficient for pressure equation
gamma=zeros(nx+2,ny+2)+1/4;
gamma(2,3:ny)=1/3;gamma(nx+1,3:ny)=1/3;gamma(3:nx,2)=1/3;gamma(3:nx,ny+1)=1/3;
gamma(2,2)=1/2;gamma(2,ny+1)=1/2;gamma(nx+1,2)=1/2;gamma(nx+1,ny+1)=1/2;

% ----------------------------------------------------------------------- %
% Solution over time
% ----------------------------------------------------------------------- %
t=0.0;
for is=1:nsteps
    
    % Boundary conditions
    u(1:nx+1,1)=2*us-u(1:nx+1,2);               % south wall
    u(1:nx+1,ny+2)=2*un-u(1:nx+1,ny+1);         % north wall
    v(1,1:ny+1)=2*vw-v(2,1:ny+1);               % west wall
    v(nx+2,1:ny+1)=2*ve-v(nx+1,1:ny+1);         % east wall
    
    % Advection-diffusion equation (predictor)
    [ut, vt] = AdvectionDiffusion2D( ut, vt, u, v, nx, ny, h, dt, nu);
    
    % Pressure equation (Poisson)
    [p, iter] = Poisson2D( p, ut, vt, gamma, nx, ny, h, dt, ...
                           beta, max_iterations, max_error );
    
    % Correction on the velocity
    u(2:nx,2:ny+1)=ut(2:nx,2:ny+1)-(dt/h)*(p(3:nx+1,2:ny+1)-p(2:nx,2:ny+1));
    v(2:nx+1,2:ny)=vt(2:nx+1,2:ny)-(dt/h)*(p(2:nx+1,3:ny+1)-p(2:nx+1,2:ny));
    
    % Print on the screen
    if (mod(is,25)==1)
        fprintf('Step: %d - Time: %f - Poisson iterations: %d\n', is, t, iter);
    end
    
    % Advance in time
    t=t+dt;
 
end

% ----------------------------------------------------------------------- %
% Final post-processing                                                   %
% ----------------------------------------------------------------------- %

% Field reconstruction
uu(1:nx+1,1:ny+1)=0.50*(u(1:nx+1,2:ny+2)+u(1:nx+1,1:ny+1));
vv(1:nx+1,1:ny+1)=0.50*(v(2:nx+2,1:ny+1)+v(1:nx+1,1:ny+1));
pp(1:nx+1,1:ny+1)=0.25*(p(1:nx+1,1:ny+1)+p(1:nx+1,2:ny+2)+...
                        p(2:nx+2,1:ny+1)+p(2:nx+2,2:ny+2));

% Surface map: u-velocity
subplot(231);
surface(X,Y,uu');
axis('square'); title('u'); xlabel('x'); ylabel('y');

% Surface map: v-velocity
subplot(234);
surface(X,Y,vv');
axis('square'); title('v'); xlabel('x'); ylabel('y');

% Surface map: pressure
% subplot(232);
% surface(X,Y,pp');
% axis('square'); title('pressure'); xlabel('x'); ylabel('y');

% Streamlines
subplot(233);
sx = 0:2*h:1;
sy = 0:2*h:1;
streamline(X,Y,uu',vv',sx,sy)
axis([0 1 0 1], 'square');
title('streamlines'); xlabel('x'); ylabel('y');

% Surface map: velocity vectors
subplot(236);
quiver(X,Y,uu',vv');
axis([0 1 0 1], 'square');
title('velocity vector field'); xlabel('x'); ylabel('y');

% Plot: velocity components along the horizontal middle axis
subplot(232);
plot(x,uu(:, round(ny/2)+1));
hold on;
plot(x,vv(:, round(ny/2)+1));
axis('square');
xlabel('x [m]');ylabel('velocity [m/s]'); title('velocity along x-axis');
legend('x-velocity', 'y-velocity');
hold off;

% Plot: velocity components along the horizontal middle axis
subplot(235);
plot(y,uu(round(nx/2)+1,:));
hold on;
plot(y,vv(round(nx/2)+1,:));
axis('square');
xlabel('y [m]');ylabel('velocity [m/s]'); title('velocity along y-axis');
legend('x-velocity', 'y-velocity');
hold off;


% ------------------------------------------------------------------- %
% Write velocity profiles along the centerlines for exp comparison
% ------------------------------------------------------------------- %
u_profile = uu(round(nx/2)+1,:);
fileVertical = fopen('experimental_data/vertical.out','w');
for i=1:ny+1 
    fprintf(fileVertical,'%f %f\n',y(i), u_profile(i));
end
fclose(fileVertical);

v_profile = vv(:,round(ny/2)+1);
fileHorizontal = fopen('experimental_data/horizontal.out','w');
for i=1:nx+1
    fprintf(fileHorizontal,'%f %f\n',x(i), v_profile(i));
end
fclose(fileHorizontal);


% ------------------------------------------------------------------- %
% Compare with exp data (available only for Re=100, 400, and 1000)
% ------------------------------------------------------------------- %
% Read experimental data from file
exp_u_along_y = dlmread('experimental_data/u_along_y.exp', '', 1, 0);
exp_v_along_x = dlmread('experimental_data/v_along_x.exp', '', 1, 0);

% Comparison with exp data
% Be careful: cols 1,2 for Re=100, 3,4 for Re=400, 5,6 for Re=1000
figure;
plot(exp_u_along_y(:,1), exp_u_along_y(:,2), 'o', y, u_profile, '-');
axis('square'); title('u along y (centerline)'); xlabel('y'); ylabel('u');

figure;
plot(exp_v_along_x(:,1), exp_v_along_x(:,2), 'o', x, v_profile, '-');
axis('square'); title('v along x (centerline)'); xlabel('x'); ylabel('v');


% --------------------------------------------------------------------------------------
% Poisson equation solver
% --------------------------------------------------------------------------------------
function [p, iter] = Poisson2D( p, ut, vt, gamma, nx, ny, h, dt, ...
                                beta, max_iterations, max_error)

    for iter=1:max_iterations
        
        for i=2:nx+1
            for j=2:ny+1
                
                delta = p(i+1,j)+p(i-1,j)+p(i,j+1)+p(i,j-1);
                S = (h/dt)*(ut(i,j)-ut(i-1,j)+vt(i,j)-vt(i,j-1));
                p(i,j)=beta*gamma(i,j)*( delta-S )+(1-beta)*p(i,j);
                
            end
        end
        
        % Estimate the error
        epsilon=0.0; 
        for i=2:nx+1
            for j=2:ny+1
                delta = p(i+1,j)+p(i-1,j)+p(i,j+1)+p(i,j-1);
                S = (h/dt)*(ut(i,j)-ut(i-1,j)+vt(i,j)-vt(i,j-1));              
                epsilon=epsilon+abs( p(i,j) - gamma(i,j)*( delta-S ) );
            end
        end
        epsilon = epsilon / (nx*ny);
        
        % Check the error
        if (epsilon <= max_error) % stop if converged
            break;
        end 
        
    end

end

% --------------------------------------------------------------------------------------
% Advection-diffusion equation
% --------------------------------------------------------------------------------------
function [ut, vt] = AdvectionDiffusion2D( ut, vt, u, v, nx, ny, h, dt, nu)
                            
    % Temporary u-velocity
    for i=2:nx
        for j=2:ny+1 
            
            ue2 = 0.25*( u(i+1,j)+u(i,j) )^2;
            uw2 = 0.25*( u(i,j)+u(i-1,j) )^2;
            unv = 0.25*( u(i,j+1)+u(i,j) )*( v(i+1,j)+v(i,j) );
            usv = 0.25*( u(i,j)+u(i,j-1) )*( v(i+1,j-1)+v(i,j-1) );
            
            A = (ue2-uw2+unv-usv)/h;
            D = (nu/h^2)*(u(i+1,j)+u(i-1,j)+u(i,j+1)+u(i,j-1)-4*u(i,j));
            
            ut(i,j)=u(i,j)+dt*(-A+D);
            
        end
    end
    
    % Temporary v-velocity
    for i=2:nx+1
        for j=2:ny 
            
            vn2 = 0.25*( v(i,j+1)+v(i,j) )^2;
            vs2 = 0.25*( v(i,j)+v(i,j-1) )^2;
            veu = 0.25*( u(i,j+1)+u(i,j) )*( v(i+1,j)+v(i,j) );
            vwu = 0.25*( u(i-1,j+1)+u(i-1,j) )*( v(i,j)+v(i-1,j) );
            A = (vn2 - vs2 + veu - vwu)/h;
            D = (nu/h^2)*(v(i+1,j)+v(i-1,j)+v(i,j+1)+v(i,j-1)-4*v(i,j));
            
            vt(i,j)=v(i,j)+dt*(-A+D);
            
        end
    end
    
end

File: codes/driven_cavity/driven_cavity_2d_staggered.cpp
// ----------------------------------------------------------------------- //
//                                    __  __  __       _  __   __          //
//      |\ /|  _  |_ |  _  |_  |__|  /   |_  |  \  _  (_ |__) | _          //
//      |   | (_| |_ | (_| |_)    |  \__ |   |__/ (_) |  |  \ |            //
//                                                                         //
// ----------------------------------------------------------------------- //
//                                                                         //
//   Author: Alberto Cuoci <alberto.cuoci@polimi.it>                       //
//   CRECK Modeling Group <http ://creckmodeling.chem.polimi.it>           //
//   Department of Chemistry, Materials and Chemical Engineering           //
//   Politecnico di Milano                                                 //
//   P.zza Leonardo da Vinci 32, 20133 Milano                              //
//                                                                         //
// ----------------------------------------------------------------------- //
//                                                                         //
//   This file is part of Matlab4CFDofRF framework.                        //
//                                                                         //
//   License                                                               //
//                                                                         //
//   Copyright(C) 2019 Alberto Cuoci                                       //
//   Matlab4CFDofRF is free software : you can redistribute it and/or      //
//   modify it under the terms of the GNU General Public License as        //
//   published by the Free Software Foundation, either version 3 of the    //
//   License, or (at your option) any later version.                       //
//                                                                         //
//   Matlab4CFDofRF is distributed in the hope that it will be useful,     //
//   but WITHOUT ANY WARRANTY; without even the implied warranty of        //
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the           //
//   GNU General Public License for more details.                          //
//                                                                         //
//   You should have received a copy of the GNU General Public License     //
//   along with Matlab4CRE.If not, see <http://www.gnu.org/licenses/>.     //
//                                                                         //
//------------------------------------------------------------------------ //
//                                                                         //
//  Code : 2D driven - cavity problem in pressure/velocity formulation     //
//        The code is adapted and extended from Tryggvason, Computational  //
//        Fluid Dynamics http ://www.nd.edu/~gtryggva/CFD-Course/          //
//                                                                         //
// ----------------------------------------------------------------------- //

#include <Eigen/Dense>
#include <Eigen/Sparse>
#include<Eigen/SparseLU>
#include <stdio.h>
#include <iostream>
#include <fstream>

// Function to write output files in Tecplot format
void WriteTecplotFile(	const std::string filename,
						const Eigen::VectorXd& x, const Eigen::VectorXd& y,
						const Eigen::MatrixXd& u, const Eigen::MatrixXd& v,
						const Eigen::MatrixXd& p);

int main()
{
	// Basic setup
	// Use only even numbers for number of cells
	const unsigned int nx = 200;						// number of cells along x
	const unsigned int ny = nx;							// number of cells along y
	const double L = 1.;								// length [m]
	const double h = 1. / static_cast<double>(nx);		// grid step along [m]
	const double nu = 1e-3;								// kinematic viscosity [m2/s]
	const double tau = 20;								// total time of simulation [s]


	// Parameters for SOR
	const unsigned int max_iterations = 10000;  // maximum number of iterations
	const double beta = 1.5;					// SOR coefficient
	const double max_error = 1e-5;				// error for convergence

	// Boundary conditions
	const double un = 1;       // north wall velocity [m/s]
	const double us = 0;       // south wall velocity [m/s]
	const double ve = 0;       // east wall velocity [m/s]
	const double vw = 0;       // west wall velocity [m/s]

	// Time step
	const double sigma = 0.5;										// safety factor for time step (stability)
	const double u2 = (un*un + us*us + ve*ve + vw*vw);				// squared velocity order of magnitude [m2/s2]
	const double dt_diff = h*h/4./nu;								// time step (diffusion stability) [s]
	const double dt_conv = 4*nu/u2;									// time step (convection stability) [s]
	const double dt = sigma*std::min(dt_diff, dt_conv);				// time step (stability) [s]
	const unsigned int nsteps = static_cast<unsigned int>(tau/dt);  // number of steps
	const double Re = un*L / nu;									// Reynolds' number

	// Output options
	const bool write_tecplot = false;

	// Summary
	std::cout << "Reynolds' number: " << Re << std::endl;
	std::cout << "Time step: " << dt << std::endl;
	std::cout << " - Diffusion: " << dt_diff << std::endl;
	std::cout << " - Convection: " << dt_conv << std::endl;

	// Memory allocation
	Eigen::MatrixXd u(nx + 1, ny + 2); u.setZero();				// x-velocity [m/s]
	Eigen::MatrixXd v(nx + 2, ny + 1); v.setZero();				// y-velocity [m/s]
	Eigen::MatrixXd p(nx + 2, ny + 2); p.setZero();				// pressure [Pa]
	Eigen::MatrixXd ut(nx + 1, ny + 2); ut.setZero();			// temporary x-velocity [m/s]
	Eigen::MatrixXd vt(nx + 2, ny + 1); vt.setZero();			// temporary y-velocity [m/s]
	Eigen::MatrixXd uu(nx + 1, ny + 1); uu.setZero();			// reconstructed x-velocity [m/s]
	Eigen::MatrixXd vv(nx + 1, ny + 1); vv.setZero();			// reconstructed y-velocity [m/s]
	Eigen::MatrixXd pp(nx + 1, ny + 1); pp.setZero();			// reconstructd pressure [Pa]

	// Grid construction
	Eigen::VectorXd x(nx+1);         // grid coordinates(x axis)
	Eigen::VectorXd y(ny+1);         // grid coordinates(y axis)
	for (unsigned int i = 0; i < nx+1; i++)
		x(i) = h*i;
	for (unsigned int j = 0; j < ny+1; j++)
		y(j) = h*j;

	// Coefficient for pressure equation
	Eigen::MatrixXd gamma(nx + 2, ny + 2); gamma.setConstant(1./4.);
	gamma(1,1) = 1./2.; 
	gamma(1,ny) = 1./2.; 
	gamma(nx,1) = 1./2.; 
	gamma(nx,ny) = 1./2.;
	for (unsigned int i = 2; i<ny; i++)
	{
		gamma(1, i) = 1./3.;
		gamma(nx, i) = 1./3.;
	}
	for (unsigned int i=2; i<nx; i++)
	{
		gamma(i, 1) = 1./3.;
		gamma(i, ny) = 1./3.;
	}

	// Time loop
	double t = 0;
	for (unsigned int istep = 1; istep <= nsteps; istep++)
	{
		// ------------------------------------------------------------------ //
		// Boundary conditions
		// ------------------------------------------------------------------ //
		for (unsigned int i = 0; i<nx+1; i++)
		{
			u(i, 0)    = 2.*us - u(i, 1);      // south wall
			u(i, ny+1) = 2.*un - u(i, ny);     // north wall
		}

		for (unsigned int i = 0; i < ny + 1; i++)
		{
			v(0, i) = 2.*vw - v(1, i);         // west wall
			v(nx+1, i) = 2.*ve - v(nx, i);     // east wall
		}

		// ------------------------------------------------------------------ //
		// Temporary velocity along x
		// ------------------------------------------------------------------ //
		for (unsigned int i=1; i<nx; i++)
			for (unsigned int j=1; j<ny+1; j++)
			{
				const double ue2 = 0.25*(u(i + 1, j) + u(i, j))*(u(i + 1, j) + u(i, j));
				const double uw2 = 0.25*(u(i, j) + u(i - 1, j))*(u(i, j) + u(i - 1, j));
				const double unv = 0.25*(u(i, j + 1) + u(i, j))*(v(i + 1, j) + v(i, j));
				const double usv = 0.25*(u(i, j) + u(i, j - 1))*(v(i + 1, j - 1) + v(i, j - 1));

				const double A = (ue2 - uw2 + unv - usv) / h;
				const double D = (nu/h/h)*(u(i + 1, j) + u(i - 1, j) + u(i, j + 1) + u(i, j - 1) - 4 * u(i, j));

				ut(i, j) = u(i, j) + dt*(-A + D);
			}

		// ------------------------------------------------------------------ //
		// Temporary velocity along x
		// ------------------------------------------------------------------ //
		for (unsigned int i = 1; i<nx+1; i++)
			for (unsigned int j = 1; j < ny; j++)
			{
				const double vn2 = 0.25*(v(i, j + 1) + v(i, j))*(v(i, j + 1) + v(i, j));
				const double vs2 = 0.25*(v(i, j) + v(i, j - 1))*(v(i, j) + v(i, j - 1));
				const double veu = 0.25*(u(i, j + 1) + u(i, j))*(v(i + 1, j) + v(i, j));
				const double vwu = 0.25*(u(i - 1, j + 1) + u(i - 1, j))*(v(i, j) + v(i - 1, j));
				const double A = (vn2 - vs2 + veu - vwu) / h;
				const double D = (nu/h/h)*(v(i + 1, j) + v(i - 1, j) + v(i, j + 1) + v(i, j - 1) - 4 * v(i, j));

				vt(i, j) = v(i, j) + dt*(-A + D);
			}

		// ------------------------------------------------------------------ //
		// Poisson equation(SOR)
		// ------------------------------------------------------------------ //
		unsigned int iter = 0;
		{
			for (iter = 1; iter <= max_iterations; iter++)
			{
				for (unsigned int i = 1; i < nx+1; i++)
					for (unsigned int j = 1; j < ny+1; j++)
					{
						const double delta = p(i + 1, j) + p(i - 1, j) + p(i, j + 1) + p(i, j - 1);
						const double S = (h / dt)*(ut(i, j) - ut(i - 1, j) + vt(i, j) - vt(i, j - 1));
						p(i, j) = beta*gamma(i, j)*(delta - S) + (1 - beta)*p(i, j);
					}

				// Estimate the error
				double epsilon = 0.0;
				for (unsigned int i = 1; i < nx + 1; i++)
					for (unsigned int j = 1; j < ny + 1; j++)
					{
						const double delta = p(i + 1, j) + p(i - 1, j) + p(i, j + 1) + p(i, j - 1);
						const double S = (h / dt)*(ut(i, j) - ut(i - 1, j) + vt(i, j) - vt(i, j - 1));
						epsilon += std::abs( p(i, j) - gamma(i, j)*(delta - S));
					}
				epsilon /= (nx*ny);

				// Check the error
				if (epsilon <= max_error) // stop if converged
					break;
			}
		}
		
		// ------------------------------------------------------------------ //
		// Correct the velocity
		// ------------------------------------------------------------------ //
		for (unsigned int i = 1; i < nx; i++)
			for (unsigned int j = 1; j < ny+1; j++)
				u(i,j) = ut(i,j) - (dt/h)*(p(i+1,j)-p(i,j));
		
		for (unsigned int i = 1; i < nx+1; i++)
			for (unsigned int j = 1; j < ny; j++)
				v(i, j) = vt(i, j) - (dt/h)*(p(i,j+1)-p(i,j));
		
		// ------------------------------------------------------------------ //
		// Interpolate velocities and pressure on the same FD grid
		// ------------------------------------------------------------------ //
		for (unsigned int i = 0; i < nx+1; i++)
			for (unsigned int j = 0; j < ny+1; j++)
			{
				uu(i,j)  = 0.5*(u(i,j+1) + u(i,j));
				vv(i,j)  = 0.5*(v(i+1,j) + v(i,j));
				pp(i, j) = 0.25*(p(i,j)+p(i+1,j)+p(i,j+1)+p(i+1,j+1));
			}
		

		// Post-processing operations
		if (istep % 10 == 1)
		{
			std::cout << "Step: " << istep << " - Time: " << t << " - Poisson iterations: " << iter << std::endl;

			if (write_tecplot == true)
			{
				std::stringstream label; label << static_cast<int>(istep/10);
				std::string filename = "Solution.tec." + label.str();
				WriteTecplotFile(filename, x, y, uu, vv, pp);
			}

		}

		// Update time step
		t += dt;
	}

	// ------------------------------------------------------------------ //
	// Write velocity profiles along the centerlines for exp comparison
	// ------------------------------------------------------------------ //
	unsigned int iaxis = static_cast<unsigned int>(nx/2);
	unsigned int jaxis = static_cast<unsigned int>(ny/2);

	std::ofstream fileVertical("vertical.txt", std::ios::out);
	fileVertical.setf(std::ios::scientific);
	for (unsigned int i = 0; i < ny+1; i++)
		fileVertical << y(i) << " " << uu(iaxis, i) << std::endl;
	fileVertical.close();

	std::ofstream fileHorizontal("horizontal.txt", std::ios::out);
	fileHorizontal.setf(std::ios::scientific);
	for (unsigned int i = 0; i < nx+1; i++)
		fileHorizontal << x(i) << " " << vv(i, jaxis) << std::endl;
	fileHorizontal.close();

	// ------------------------------------------------------------------ //
	// Write Tecplot file
	// ------------------------------------------------------------------ //
	WriteTecplotFile("SolutionFinal.tec", x, y, uu, vv, pp);

	std::cout << "Calculations completed" << std::endl;
	std::cout << "Press enter to exit..." << std::endl;
	getchar();
	
	return 0;
}

void WriteTecplotFile(	const std::string filename, 
						const Eigen::VectorXd& x, const Eigen::VectorXd& y,
						const Eigen::MatrixXd& u, const Eigen::MatrixXd& v,
						const Eigen::MatrixXd& p)
{
	std::ofstream fTecplot(filename.c_str(), std::ios::out);
	fTecplot.setf(std::ios::scientific);
	fTecplot << "Title = Solution" << std::endl;
	fTecplot << "Variables = \"x\", \"y\", \"u\", \"v\", \"p\" " << std::endl;
	fTecplot << "Zone I = " << x.size() << ", J = " << y.size() << ", F = POINT" << std::endl;

	for (unsigned int i = 0; i < x.size(); i++)
		for (unsigned int j = 0; j < y.size(); j++)
			fTecplot << x(i) << " " << y(j) << " " << u(i, j) << " " << v(i, j) << " " << p(i, j) << std::endl;
	fTecplot.close();
}
